### PR TITLE
Auto-infer output_size for padding="valid" in keras.ops.image.reconstruct_patches

### PR DESCRIPTION
**Pull request by brianmanderson**
## Description
Third PR in the `reconstruct_patches` series (follow-up issue: #23318),
following #22984 (op) and #23267 (`channels_first` support).

With `padding="valid"` and non-overlapping patches, the reconstructed region
is fully determined by the patch grid: `grid * size` per spatial dim. This PR
makes `output_size` optional in that case:  it defaults to `None` and is
inferred from the patch shape, so the common round-trip no longer requires
carrying the original size around:

```python
patches = keras.ops.image.extract_patches(images, size=(5, 5))
recon = keras.ops.image.reconstruct_patches(patches, size=(5, 5))
```

Details:

- Inference works in both the eager and symbolic paths and for both
  `data_format`s. At trace time, unknown (dynamic) grid dims produce an
  unknown symbolic spatial dim; on the eager path they raise a clear
  `ValueError` asking for an explicit `output_size`.
- When `output_size` *is* given with `padding="valid"`, it is now validated
  against the static grid dims (`output_size == grid * size`), turning silent
  shape mistakes into immediate errors.
- `padding="same"` still requires `output_size` (the original size is
  ambiguous from patches alone). This precondition is now also checked at
  graph-construction time in the `Operation.__init__`, not just on the first
  eager call — so a functional model can no longer be built that only fails
  once data flows through it.
- Backward compatible: `output_size` keeps its positional slot (now with
  default `None`), so all existing call sites are unaffected.

Tests cover eager + symbolic inference for 2D/3D, both data formats, the
`valid` output-size validation, dynamic-dim behavior, and the earlier
`same`-without-`output_size` error, on all backends.

## Contributor Agreement

**Please review our [AI-Assisted Contribution Policy](../CONTRIBUTING.md#ai-assisted-contribution-policy) and check all boxes below before submitting your PR for review:**

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

**Note:** Failing to adhere to this agreement may result in your future PRs no longer being reviewed.

### CODE REVIEW

File: keras/src/ops/image.py [hunk 1]
```diff
@@ -1013,25 +1019,41 @@ def compute_output_spec(self, patches):
...
+        # the grid — `__init__` guarantees `padding="valid"` when
+        # `output_size` is None. Unknown grid dims stay None.
+        if self.output_size is None:
+            if any(g is None for g in grid):
+                spatial = (None,) * len(self.size)
+            else:
+                spatial = tuple(
+                    (g - 1) * s + k
+                    for g, s, k in zip(grid, self.strides, self.size)
+                )
```
[the remark this answers is not in the document]
Comment (brianmanderson) [reply keras/src/ops/image.py]: Done. compute_output_spec now resolves each spatial dim individually, so mixed static/dynamic grids keep their static dims, and uses isinstance(g, int) per the style guide. Added a test for the mixed case ((2, None, 4, 75) → (2, None, 20, 3)).

File: keras/src/ops/image.py [hunk 2]
```diff
@@ -1199,6 +1223,30 @@ def _validate_reconstruct_strides(size, strides, fn_name):
...
+        grid_start = 0 if rank == n + 1 else 1
+    else:
+        grid_start = 1 if rank == n + 1 else 2
+    grid = patches.shape[grid_start : grid_start + n]
+    if any(g is None for g in grid):
+        raise ValueError(
+            "Cannot auto-infer `output_size` for `padding='valid'`: at "
+            "least one patch-grid dimension is unknown "
+            f"(patches.shape={patches.shape}). Pass `output_size` explicitly."
+        )
```
[the remark this answers is not in the document]
Comment (brianmanderson) [reply keras/src/ops/image.py]: Done. Switched to isinstance(g, int).

File: keras/src/ops/image.py [hunk 3]
```diff
@@ -1013,25 +1019,41 @@ def compute_output_spec(self, patches):
...
+                if not isinstance(g, int):
+                    continue
+                if self.padding == "valid":
+                    if g * p != o:
+                        raise ValueError(
+                            f"`padding='valid'` requires output_size to "
+                            f"equal size * grid. Got output_size="
+                            f"{self.output_size}, grid {dim_name}={g}, "
+                            f"size={self.size}."
+                        )
```
[the remark this answers is not in the document]
Comment (brianmanderson) [reply keras/src/ops/image.py]: Done. The symbolic validation now uses (g - 1) * s + p (equivalent while strides == size is enforced, but consistent with the inference formula and ready for the overlap follow-up). Also applied the same formula to the corresponding eager-path checks in the 2D/3D helpers and updated the docstring so all three agree.

File: keras/src/ops/image.py [hunk 4]
```diff
@@ -1214,17 +1262,29 @@ def _reconstruct_patches_2d(
...
         )
-    if len(output_size) != 2:
-        raise ValueError(
-            "Invalid `output_size`. Expected length 2 (H, W). "
-            f"Got: output_size={output_size}"
-        )
     if padding not in ("same", "valid"):
         raise ValueError(
             f"Invalid `padding`. Expected 'same' or 'valid'. Got: {padding}"
         )
```
[the remark this answers is not in the document]
Comment (brianmanderson) [reply keras/src/ops/image.py:1266]: Done. Rank check now runs at the top of _reconstruct_patches_2d, before output_size inference can see a malformed shape. The padding validation was already present just below that point. The now-redundant downstream rank checks were removed.

File: keras/src/ops/image.py [hunk 5]
```diff
@@ -1324,17 +1384,29 @@ def _reconstruct_patches_3d(
...
 ):
-    if len(output_size) != 3:
-        raise ValueError(
-            "Invalid `output_size`. Expected length 3 (D, H, W). "
-            f"Got: output_size={output_size}"
-        )
     if padding not in ("same", "valid"):
         raise ValueError(
             f"Invalid `padding`. Expected 'same' or 'valid'. Got: {padding}"
         )
```
[the remark this answers is not in the document]
Comment (brianmanderson) [reply keras/src/ops/image.py:1383]: Done. Same treatment as the 2D path.

**Comment by brianmanderson** [comment]
@keerthanakadiri hope this finds you well! Just wanted to reach out and ask if there is anything I can provide to help with this merge, thank you!
@hertschuh adding you as well with this question, thank you!

**Review by hertschuh — APPROVED** [review]
Thanks!

**Comment by brianmanderson** [comment]
Thank you!
